File: 7/b.ts
// const input = await Bun.file("test.txt").text();
const input = await Bun.file("input.txt").text();

let res = 0;

const cardStrengths = [
  "A",
  "K",
  "Q",
  "T",
  "9",
  "8",
  "7",
  "6",
  "5",
  "4",
  "3",
  "2",
  "J",
];

const getType = (cards: string[]) => {
  // get max number of same cards
  let max = 0;
  let maxCard = "";
  for (let i = 0; i < cards.length; i++) {
    let card = cards[i];
    let count = cards.filter((d) => d === card || d === "J").length;
    if (count > max) {
      max = count;
      maxCard = card;
    }
  }
  const otherCards = cards.filter((d) => d !== maxCard && d !== "J");
  if (max === 5) {
    return 8;
  }
  if (max === 4) {
    return 7;
  }
  if (max === 3 && otherCards[0] === otherCards[1]) {
    return 6;
  }
  if (max === 3) {
    return 3;
  }
  if (
    max === 2 &&
    (otherCards[0] === otherCards[1] ||
      otherCards[0] === otherCards[2] ||
      otherCards[1] === otherCards[2])
  ) {
    return 2;
  }
  if (max === 2) {
    return 1;
  }
  return 0;
};

const lines = input.split("\n").map((d) => {
  let a = d.split(" ");
  let cards = a[0].split("");
  let type = getType(cards);
  console.log(cards, type);
  return {
    cards,
    bid: parseInt(a[1]),
    type: type,
  };
});

lines.sort((a, b) => {
  let typeA = a.type;
  let typeB = b.type;
  if (typeA !== typeB) {
    return typeA - typeB;
  }
  let i = 0;
  while (i < a.cards.length) {
    let cardA = a.cards[i];
    let cardB = b.cards[i];
    let indexA = cardStrengths.indexOf(cardA);
    let indexB = cardStrengths.indexOf(cardB);
    if (indexA !== indexB) {
      return indexB - indexA;
    }
    i++;
  }
  return 0;
});

// console.log(lines);

lines.forEach((d, i) => {
  console.log(i + 1, d.bid);
  res += d.bid * (i + 1);
});

console.log(res);

File: 7/b.go
package main

import (
	"fmt"
	"os"
	"sort"
	"strconv"
	"strings"
)

type Hand struct {
	cards       string
	bid         int
	combination int
}

func main() {
	file, _ := os.ReadFile("input.txt")

	res := 0

	lines := strings.Split(string(file), "\n")

	hands := make([]Hand, len(lines))

	cardStrengths := "J23456789TQKA"
	for i, line := range lines {
		rawCards, rawBid, _ := strings.Cut(line, " ")
		cardCount := make(map[rune]int)
		jockers := 0
		combination := 0
		for _, card := range rawCards {
			if card == 'J' {
				jockers++
				continue
			}
			cardCount[card]++
		}
		counts := make([]int, 1, len(cardCount)+1)
		for _, count := range cardCount {
			counts = append(counts, count)
		}
		sort.Ints(counts)
		counts[len(counts)-1] += jockers
		for _, count := range counts {
			combination += count * count
		}

		bid, _ := strconv.Atoi(rawBid)

		hands[i] = Hand{rawCards, bid, combination}
	}

	sort.Slice(hands, func(i, j int) bool {
		if hands[i].combination != hands[j].combination {
			return hands[i].combination < hands[j].combination
		}
		for k := 0; k < 5; k++ {
			if hands[i].cards[k] != hands[j].cards[k] {
				indexI := strings.IndexRune(cardStrengths, rune(hands[i].cards[k]))
				indexJ := strings.IndexRune(cardStrengths, rune(hands[j].cards[k]))
				return indexI < indexJ
			}
		}
		return false
	})

	for i, hand := range hands {
		res += hand.bid * (i + 1)
	}

	fmt.Println(res)
}
